daemon/policy.go

```diff
@@ -266,7 +289,10 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 
 	logger.WithField(logfields.PolicyRevision, rev).Info("Policy imported via API, recalculating...")
 
-	d.TriggerPolicyUpdates(false, "policy rules added")
+	// Only regenerate endpoints which are needed to be regenerated as a result
+	// of the rule update. The rules which were imported most likely do not select
+	// all endpoints in the policy repository (and may not select any at all).
+	go d.ReactToRuleUpdates(&policySelectionWG, epsIdentityConsumers, endpointsToRegen, rev)
```

Comment:
We have three level of go routines beyond this point:

1. go d.ReactToRuleUpdates()
   1.1 Wait for waitgroup
   1.2 go bumpEndpointRevisions
      1.2.1 go ep.PolicyRevisionBumpEvent(rev)
   1.3 endpointmanager.RegenerateEndpointSet()

If we have multiple calls to PolicyAdd(), what is guarnateeing the order to be correct in the event queue? Assuming that two PolicyAdd() calls are made, both schedule a go routine to wait for the list to be calculated. Both complete roughly at the same time, the order on which either of them could follow up and enqueue the event is random.

Reply:
This should be resolved with the introduction of the new queues for the policy repository. Let me know what you think.